File: 2022/day_2/src/main.rs
#![allow(dead_code, unused_variables, unused_mut)]

use aoc_std::*;

const SHAPE_SCORE: [(&str, isize); 3] = [("X", 1), ("Y", 2), ("Z", 3)];
const OPPONENT_SCORE: [(&str, isize); 3] = [("A", 1), ("B", 2), ("C", 3)];
const OUTCOME_SCORE: [(&str, isize); 3] = [("X", 0), ("Y", 3), ("Z", 6)];

fn get_score(key: &str, map: &[(&str, isize)]) -> isize {
    map.iter()
        .find(|&(k, _)| k == &key)
        .map(|&(_, v)| v)
        .unwrap()
}

fn did_win(split: &Vec<&str>) -> usize {
    let diff = get_score(split[0], &OPPONENT_SCORE) - get_score(split[1], &SHAPE_SCORE);
    if split == &vec!["C", "X"] {
        return 6;
    }
    match diff {
        0 => 3,
        -1 => 6,
        _ => 0,
    }
}

fn part1(file_lines: &Vec<String>) -> usize {
    let mut score: usize = 0;
    for line in file_lines {
        let split = line.split(" ").collect::<Vec<&str>>();
        score += get_score(split[1], &SHAPE_SCORE) as usize + did_win(&split);
    }
    score
}

fn choice_score(split: &Vec<&str>) -> usize {
    if split[1] == "X" {
        let tmp = get_score(split[0], &OPPONENT_SCORE) as usize + 2;
        return if tmp > 3 { tmp - 3 } else { tmp };
    } else if split[1] == "Y" {
        return get_score(split[0], &OPPONENT_SCORE) as usize;
    }
    let tmp = get_score(split[0], &OPPONENT_SCORE) as usize + 1;
    return if tmp > 3 { tmp - 3 } else { tmp };
}

fn part2(file_lines: &Vec<String>) -> usize {
    let mut score: usize = 0;
    for line in file_lines {
        let split = line.split(" ").collect::<Vec<&str>>();
        score += choice_score(&split) + get_score(split[1], &OUTCOME_SCORE) as usize;
    }
    score
}

fn main() {
    let file_lines = load_file_lines("input.txt");

    println!("[Part 1] Answer: {}", part1(&file_lines));
    println!("[Part 2] Answer: {}", part2(&file_lines));
}

File: 2022/day_2/main.py
shape_score = {"X": 1, "Y": 2, "Z": 3}
opponent_score = {"A": 1, "B": 2, "C": 3}
outcome_score = {"X": 0, "Y": 3, "Z": 6}


def did_win(split):
    diff = opponent_score[split[0]] - shape_score[split[1]]
    return 3 if diff == 0 else (6 if diff == -1 or split == ["C", "X"] else 0)


def choice_score(split):
    if split[1] == "X":  # loose
        tmp = opponent_score[split[0]] + 2
        return tmp - 3 if tmp > 3 else tmp
    elif split[1] == "Y":  # draw
        return opponent_score[split[0]]
    tmp = opponent_score[split[0]] + 1  # win
    return tmp - 3 if tmp > 3 else tmp


part1, part2 = 0, 0
for line in open("./input.txt").readlines():
    split = line.strip().split()
    part1 += shape_score[split[1]] + did_win(split)
    part2 += choice_score(split) + outcome_score[split[1]]
print(f"Part 1 answer: {part1}\nPart 2 answer: {part2}")
